collections/_article/introducing-reverse-z.md

```diff
@@ -0,0 +1,135 @@
+---
+title: "Introducing Reverse Z (AKA I'm sorry for breaking your shader)"
+excerpt: "We are breaking compatibility for some custom shaders. Here is why."
+categories: ["news"]
+author: Clay John
+image: /storage/blog/covers/reverse-z.webp
+date: 2024-04-19 15:00:00
+---
+
+After extensive discussion, we have decided to implement the reverse z depth buffer technique in Godot 4.3. This is an exciting change as it brings a massive improvement to depth buffer precision at no performance or memory cost. This technique is used everywhere in 3D games these days. In practical terms, it significantly reduces the chances of running into z-fighting and other depth buffer precision artifacts. NVIDIA has an [excellent article](https://developer.nvidia.com/content/depth-precision-visualized) explaining the theory and benefits behind using reverse z, please read it for more technical info.
+
+I am writing this post because, unfortunately, implementing reverse z naturally breaks compatibility for some shaders. We try to avoid compatibility breakage as much as possible, but in some cases it is unavoidable, or the benefits of doing so far outweigh the cost. The rendering team felt in this case that the benefits sufficiently outweighed the costs.
+
+We are certain that the vast majority of users will not run into any compatibility breakage. For most people this change is all good with no downside. 
+
+You may need to tweak your shaders if you use a custom spatial Shader that:
+
+- Writes to ``POSITION`` in the vertex processor function
+
+- Writes to ``DEPTH``  in the fragment processor function
+
+- Reads from depth_texture
+
+- Operates in clip space
+
+In most cases when working with ``POSITION``, ``DEPTH``, or the depth_texture, you won't need to make any shader changes as long as you are transforming the values into/out of clip space using the ``PROJECTION_MATRIX`` or the ``INV_PROJECTION_MATRIX``. In those cases, the transformation is handled for you and your shader will continue to work. 
+
+Let's look at the cases one-by-one to see what sort of shaders will break and how to fix them.
+
+
+### Writes to ``POSITION``
+
+When you write to ``POSITION`` in Godot, you are bypassing the built in vertex transformation and writing a clip space position directly. This can be helpful to optimize a vertex shader, or to achieve certain effects (like having a mesh stay fixed in the camera view). Commonly users wrote the following line:
```

Comment:
```suggestion
When you write to ``POSITION`` in Godot, you are bypassing the built-in vertex transformation and writing a clip space position directly. This can be helpful to optimize a vertex shader, or to achieve certain effects (like having a mesh stay fixed in the camera view). Commonly users wrote the following line:
```